src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,335 @@
+    s 
+        One-dimensional |NumPy array| of the approximated singular values
+    Vh
+        |VectorArray| of the approximated right singular vectors
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+    
+    assert isinstance(A, Operator)
+    
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
+    assert 0 <= p <= max(A.source.dim, A.range.dim) - modes and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
+
+    if range_product is None:
+        range_product = IdentityOperator(A.range)
+    else:
+        assert isinstance(range_product, Operator)
+        assert range_product.source == range_product.range == A.range
+        
+    if source_product is None:
+        source_product = IdentityOperator(A.source)
+    else:
+        assert isinstance(source_product, Operator)
+        assert source_product.source == source_product.range == A.source
+        
+    if A.source.dim == 0 or A.range.dim == 0:
+        return A.source.empty(), np.array([]), A.range.empty()
+    
+    Q = rrf(A, source_product=source_product, range_product=range_product, q=q, l=modes+p)
+    B = A.apply_adjoint(range_product.apply(Q))
+    Q_B, R_B = gram_schmidt(source_product.apply_inverse(B), product=source_product, return_R=True)
+    U_b, s, Vh_b = sp.linalg.svd(R_B.T, full_matrices=False) 
+
+    with logger.block(f'Computing generalized left-singular vectors ({modes} vectors) ...'):
+        U = Q.lincomb(U_b.T)
+
+    with logger.block(f'Computung generalized right-singular vector ({modes} vectors) ...'):
+        Vh = Q_B.lincomb(Vh_b)
+        
+    return  U[:modes], s[:modes], Vh[:modes]
+
+
+@defaults('modes', 'p', 'q')
+def random_ghep(A, E=None, modes=6, p=20, q=2, single_pass=False):
+    r"""Approximates a few eigenvalues of a symmetric linear |Operator| with randomized methods.
+
+    Approximates `modes` eigenvalues `w` with corresponding eigenvectors `v` which solve
+    the eigenvalue problem.
+    
+    .. math::
+        A v_i = w_i v_i
+
+    or the generalized eigenvalue problem
+
+    .. math::
+        A v_i = w_i E v_i
+
+    if `E` is not `None`.
+
+    This method is an implementation of algorithm 6 and 7 in :cite:`SJK21`. 
+
+    Parameters
+    ----------
+    A
+        The hermitian linear |Operator| for which the eigenvalues are to be computed.
+    E
+        The hermitian |Operator| which defines the generalized eigenvalue problem.
+    modes
+        The number of eigenvalues and eigenvectors which are to be computed.
+    p :
+        If not `0`, adds `p` colums to the randomly sampled matrix in the :func:`rrf` method 
+        (oversampling parameter).
+    q : 
+        If not `0`, performs `q` power iterations to increase the relative weight
+        of the larger singular values.
+    single_pass
+        If `True`, computes the ghep where only one set of matvec Ax is required, but an additional error is produced. 
```

Comment:
'but an additional error is produced' sounds strange. Maybe 'at the expense of lower numerical accuracy' or so?